File: MyDailyBeat/MyDailyBeat/EVCBankViewController.m
//
//  EVCBankViewController.m
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 5/28/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//

#import "EVCBankViewController.h"
#import "Constants.h"
#import "VerveBankObject.h"
#import "RestAPI.h"
#import "EVCCommonMethods.h"

@interface EVCBankViewController ()

@end

@implementation EVCBankViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    [self retrieveBankData];
}

- (void) viewDidAppear:(BOOL)animated {
    NSData *data = [[NSUserDefaults standardUserDefaults] objectForKey:@"myBank"];
    BankInfo * temp = [NSKeyedUnarchiver unarchiveObjectWithData:data];
    if (temp != nil) {
        self.bank = temp;
        [self retrieveBankData];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (id) initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil andBank:(BankInfo *) bankNotNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.bank = bankNotNil;
    }
    return self;
}

- (IBAction)gotoBank:(id)sender {
    if ([self isAppInstalled:self.bank.appName]) {
        NSString *name = self.bank.appName;
        name = [name stringByReplacingOccurrencesOfString:@" " withString:@"-"];
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:[name stringByAppendingString:@"://"]]];
    } else {
        NSLog(@"%@", self.bank);
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:self.bank.appURL]];
    }
}

- (BOOL) isAppInstalled: (NSString *) name {
    name = [name stringByReplacingOccurrencesOfString:@" " withString:@"-"];
    NSLog(@"%@", [name stringByAppendingString:@"://"]);
    return [[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:[name stringByAppendingString:@"://"]]];
}

- (BOOL) doesAppExist: (NSString *) name {
    __block BOOL val = false;
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        val = [[RestAPI getInstance] doesAppExistWithTerm:name andCountry:@"US"];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
        });
    });
    
    return val;
    
}

- (void) retrieveBankData {
    if (self.bank != nil) {
        [nilLabel setHidden:YES];
        [imgView setHidden:NO];
        [btn setHidden:NO];
        [bankNameLbl setHidden:NO];
        dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
        dispatch_async(queue, ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view makeToastActivity];
            });
            NSURL *imgurl = [NSURL URLWithString:self.bank.iconURL];
            NSData *data = [NSData dataWithContentsOfURL:imgurl];
            UIImage *img = [[UIImage alloc] initWithData:data];
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view hideToastActivity];
                [imgView setImage:[EVCCommonMethods imageWithImage:img scaledToSize:CGSizeMake(120, 120)]];
                [bankNameLbl setText:self.bank.appName];
            });
        });
    } else {
        [nilLabel setHidden:NO];
        [imgView setHidden:YES];
        [btn setHidden:YES];
        [bankNameLbl setHidden:YES];
        [nilLabel setText:@"My Bank not Set"];
    }
}


@end

File: MDB_Swift/MyDailyBeat/MyDailyBeat/src/EVCBankViewController.swift

//
//  EVCBankViewController.swift
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 5/28/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//
import UIKit
import Toast_Swift
import API
class EVCBankViewController: UIViewController {

    @IBOutlet var imgView: UIImageView!
    @IBOutlet var btn: UIButton!
    @IBOutlet var bankNameLbl: UILabel!
    @IBOutlet var nilLabel: UILabel!
    var bank: BankInfo!

    @IBAction func gotoBank(_ sender: Any) {
        let appURL: String = self.bank.appURL
        UIApplication.shared.open(URL(string: appURL)!, options: [:], completionHandler: nil)
        
    }


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view from its nib.
        self.retrieveBankData()
    }

    override func viewDidAppear(_ animated: Bool) {
        var data: Data? = UserDefaults.standard.object(forKey: "myBank") as! Data?
        if let bankData = data, let temp =  NSKeyedUnarchiver.unarchiveObject(with: bankData) as? BankInfo {
            self.bank = temp
            self.retrieveBankData()
        }
    }

    

    func isAppInstalled(_ name: String) -> Bool {
        var realname = name.replacingOccurrences(of: " ", with: "-")
        return UIApplication.shared.canOpenURL(URL(string: realname + "://")!)
    }

    func doesAppExist(_ name: String) -> Bool {
        var val: Bool = false
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
            })
            val = RestAPI.getInstance().doesAppExist(withTerm: name, andCountry: "US")
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
            })
        })
        return val
    }

    func retrieveBankData() {
        if self.bank != nil {
            nilLabel.isHidden = true
            imgView.isHidden = false
            btn.isHidden = false
            bankNameLbl.isHidden = false
            
            DispatchQueue.global().async(execute: {() -> Void in
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.makeToastActivity(ToastPosition.center)
                })
                var imgurl = URL(string: self.bank.iconURL)
                var data: Data? = RestAPI.getInstance().fetchImage(atRemoteURL: imgurl!)
                var img = UIImage(data: data!)
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.hideToastActivity()
                    self.imgView.image = EVCCommonMethods.image(with: img!, scaledTo: CGSize(width: CGFloat(120), height: CGFloat(120)))
                    self.bankNameLbl.text = self.bank.appName
                })
            })
        }
        else {
            nilLabel.isHidden = false
            imgView.isHidden = true
            btn.isHidden = true
            bankNameLbl.isHidden = true
            nilLabel.text = "My Bank not Set"
        }
    }
}
